### ISSUE TITLE
Strategy and RuleT design issues

### ISSUE THREAD

**Issue by fpoli** [fpoli/viper-silver#120]
Some rewriters have unexpected results when updating both the node and the context. As an example, consider the following one intended to eliminate _Not_ nodes by negating its containing expression. It would rewrite _Not\(False\)_ into _True_, for example.

‌

```scala
{
  case (n: Not, c) =>
    (n.exp, c.copy(negate = true))
    
  case (n, c) if c.negate =>
    // Negate expression, i. e., true and false literals, apply De Morgan, ...
}
```

‌

Although this top-down rewriter is straightforward, it doesn’t work. Analyzing the issue carefully, it turns out that the current design applies the function above but trashes the returned context, only keeping the rewritten node. It does so by invoking the _RuleT.execute_ method.

Later on, _Context.updateFunc_ is invoked in order to update the context only. The same function above is invoked one second time, trashing the returned node now and keeping the context. Only that, by this time, it won’t match the same case, as the _Not_ node is no longer there and therefore the returned context on the second run differs from the one returned previously.

The solution is to consider that rewriters always rewrites nodes and contexts at once. In other words, _RuleT.execute_ should produce the tuple _\(N, C\)_ always. Today, this method is retuning _Option\[N\]_. Notice that the proposed signature also eliminates _Option_ as most of the client code systematically test if the result is _None_ and return the original node \(and context\) in that case. This will significantly simplifies the strategy combinators and clients in general.

Additionally the present issue could also consider to eliminate ViperStrategy, as Strategy seems to provide all the features there and ViperStrategy contains copied code from Strategy. It would also make sense to try to eliminate _Context.updateFunc_ mentioned above, should its functionality be completely provided by the proposed refactoring.

This issue is a prerequisite to #286.

**Comment by fpoli** [comment]
> **@caterinaurban** commented on 2016-03-07 17:05

**Comment by fpoli** [comment]
> **@meilers** commented on 2016-03-29 11:38

**Comment by fpoli** [comment]
> **@meilers** commented on 2016-03-29 11:38

Fixed in commit cabb732.
